client/src/main/java/io/pravega/client/batch/impl/BatchClientImpl.java

```diff
@@ -60,43 +66,85 @@ private StreamInfo getStreamInfo(Stream stream) {
     }
 
     @Override
-    public Iterator<SegmentInfo> listSegments(Stream stream) {
-        return listSegments(stream, new Date(0L));
+    public StreamSegmentInfo getSegments(Stream stream) {
+        return listSegments(stream, Optional.empty(), Optional.empty());
     }
 
-    private Iterator<SegmentInfo> listSegments(Stream stream, Date from) {
-        // modify iteration above but starting with a timestamp and ending with a break
-        Map<Segment, Long> segments = getAndHandleExceptions(controller.getSegmentsAtTime(new StreamImpl(stream.getScope(),
-                                                                                                         stream.getStreamName()),
-                                                                                          from.getTime()),
-                                                             RuntimeException::new);
+    @Override
+    public StreamSegmentInfo getSegments(final Stream stream, StreamCut fromStreamCut, StreamCut toStreamCut) {
+        Preconditions.checkNotNull(stream, "stream");
+        return listSegments(stream, Optional.ofNullable(fromStreamCut), Optional.ofNullable(toStreamCut));
+    }
+
+    @Override
+    public <T> SegmentIterator<T> readSegment(final SegmentMetadata segment, final Serializer<T> deserializer) {
+        return new SegmentIteratorImpl<>(inputStreamFactory, segment.getSegment(), deserializer,
+                segment.getStartOffset(), segment.getEndOffset());
+    }
+
+    private StreamSegmentInfo listSegments(final Stream stream, final Optional<StreamCut> startStreamCut,
+                                           final Optional<StreamCut> toStreamCut) {
+        //Validate that the stream cuts are for the stream requested stream.
+        startStreamCut.ifPresent(streamCut -> Preconditions.checkArgument(stream.equals(streamCut.getStream())));
+        toStreamCut.ifPresent(streamCut -> Preconditions.checkArgument(stream.equals(streamCut.getStream())));
+
+        // if startStreamCut is not provided use the streamCut at the start of the stream.
+        // if toStreamCut is not provided obtain a streamCut at the tail of the stream.
+        return getStreamSegmentInfo(startStreamCut.orElse(fetchStreamCut(stream, new Date(0L))),
+                toStreamCut.orElse(fetchTailStreamCut(stream)));
+    }
+
+    private StreamCut fetchStreamCut(final Stream stream, final Date from) {
+        Map<Segment, Long> segments = getAndHandleExceptions(controller.getSegmentsAtTime(
+                new StreamImpl(stream.getScope(), stream.getStreamName()), from.getTime()), RuntimeException::new);
+        return new StreamCut(stream, segments);
+    }
+
+    private StreamCut fetchTailStreamCut(final Stream stream) {
+        StreamSegments currentSegments = getAndHandleExceptions(controller
```

Comment:
you can also chain the next block (getSegmentInfo) in the future for controller.currentSegments and then do a get on it overall.

Reply:
done.